docs/adr/0004-only-threadsafe-interfaces.md

```diff
@@ -0,0 +1,156 @@
+# Remove support for non-`Send+Sync` interfaces
+
+* Status: proposed
+* Deciders: mhammond, rfkelly, travis; proposed: jhugman, dmose, janerik
+* Date: 2021-03-31
+
+Technical Story: [Issue 419](https://github.com/mozilla/uniffi-rs/issues/419)
+
+## Context and Problem Statement
+
+[ADR-0003](0003-threadsafe-interfaces.md) introduced support for "thread-safe
+interfaces" - possibly leading to the impression that there is such a thing as
+non-threadsafe interfaces and confusion about exactly what the attribute means.
+
+However, the entire concept of non-threadsafe interfaces is a mismomer -
+everything wrapped by uniffi is thread-safe - the only question is who manages
+this thread-safety. Interfaces which are not marked as thread-safe cause uniffi
+to wrap the interface in a mutex which is hidden in the generated code and
```

Comment:
No, I think it's accurate as written. If it's not marked as "threadsafe" then we wrap it in a mutex, which is threadsafe. If it is marked as [threadsafe] but is not Send+Sync it will not compile.

I'll clarify this though by changing that line to "The Rust compiler insists that everything wrapped by uniffi is thread-safe."

Reply:
Ah, that's great.  Thanks!